### DISCUSSION TITLE
How does theia-cloud dispose container pod?

### DISCUSSION THREAD

**Discussion by inlann** [eclipse-theia/theia-cloud#263] [Q&A] [answered]
Hello:

I deployed theia-cloud in a minikube container and found that the IDE container would be destroyed 30 minutes after it was no longer being used, even if the webpage of the IDE container was not closed. However, when I tried to replace the theia-cloud-demo image with my own image built with theia version 1.37 and successfully started it, the container did not automatically destroy itself even after I actively closed all the webpages. I tried to find out the logic in theia-cloud for determining whether an IDE container is still in use and for its automatic destruction, but the architectural documentation of theia-cloud only defines a session's timeout time in the [App Definition CR](https://github.com/eclipsesource/theia-cloud/blob/main/doc/docs/Architecture.md#app-definition). In the [AppDefinitionSpec.java](https://github.com/eclipsesource/theia-cloud/blob/main/java/common/org.eclipse.theia.cloud.common/src/main/java/org/eclipse/theia/cloud/common/k8s/resource/AppDefinitionSpec.java), I saw the [Timeout](https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/java/common/org.eclipse.theia.cloud.common/src/main/java/org/eclipse/theia/cloud/common/k8s/resource/AppDefinitionSpec.java#L161C25-L161C32) class and the [Monitor](https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/java/common/org.eclipse.theia.cloud.common/src/main/java/org/eclipse/theia/cloud/common/k8s/resource/AppDefinitionSpec.java#L190) class, are these related to detection? Can you explain how theia-cloud determines whether an IDE container is still in use, and is it through the network traffic passed to the IDE container?

Really appreciate!

**Comment by inlann** [comment]
Does the [monitor-theia](https://github.com/eclipsesource/theia-cloud/tree/main/node/monitor-theia) is used to monitor the usage of theia ide container?

**Reply by sgraband** [reply]
Hey @inlann,

there are currently 2 mechanisms to decide whether or not to shutdown a pod (note that you can use both at the same time):

- a fixed timeout after X minutes, which is configured via the [AppDefinition CRs](https://github.com/eclipsesource/theia-cloud/blob/main/doc/docs/Architecture.md#app-definition) timeout object. If this is not provided the pod can potentially run forever.
- the second option is using a monitor service. Here you can define a number in minutes of inactivity after which the user is asked about, whether he is still there (notifiyAfter) and after how many minutes of inactivity you want to shutdown the pod (timeoutAfter). You also need to provide the port information of where the monitor service is running inside of the pod.

The service running inside the pod is either a [vscode-extension](https://github.com/eclipsesource/theia-cloud/tree/main/node/monitor) or a [Theia extension](https://github.com/eclipsesource/theia-cloud/tree/main/node/monitor-theia) installed in the application. You can take a look at the [demo applications](https://github.com/eclipsesource/theia-cloud/tree/main/demo/dockerfiles) to see how this is done for the theia-cloud-demo.

For what exactly is considered an activity you can take a look at the two extenions. The implementation is slightly different between the two due to the different capabilities that vscode and Theia extensions have.

The behavior is of course customizable and you can provide your own extension that reports whatever you find suitable as an activity.

Hope this helps.

**Reply by inlann** [reply]
Hi @sgraband,

Really appreciate for you explanation!

I reviewed all the source code of two extensions, monitor (a VSCode extension) and theia-monitor (a Theia extension). Both plugins listen to the mouse and keyboard events of users in their respective applications (VSCode and Theia) to determine if the application is still in use. 

Taking theia-monitor as an example, its backend module for the Theia backend process, namely the ExpressJS server, adds a GET request route: `/monitor/activity/lastActivity` and two POST request routes: `/monitor/message` and `/monitor/activity/popup`, to provide external access: 

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/node/monitor-theia/src/node/monitor-rest-service-contribution.ts#L16-L18

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/node/monitor-theia/src/node/modules/activity-tracker-backend-module-impl.ts#L20-L51

The frontend module of theia-monitor adds three window listeners - keydown, mousedown, and mousemove - upon initialization:

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/node/monitor-theia/src/browser/modules/activity-tracker-frontend-module-impl.ts#L58-L60

This means that mouse and keyboard events in the Theia window are recorded by theia-monitor with timestamps in its own state. 

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/node/monitor-theia/src/node/modules/activity-tracker-backend-module-impl.ts#L53-L61

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/node/monitor-theia/src/node/modules/activity-tracker-backend-module-impl.ts#L15-L17

When an external GET request accesses `/monitor/activity/lastActivity`, the backend of theia-monitor returns this timestamp, which represents the time point of the user's last action:

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/node/monitor-theia/src/node/modules/activity-tracker-backend-module-impl.ts#L23-L27

I noticed that Java's operator module accesses this theia-monitor's `/monitor/activity/lastActivity` interface of the Theia IDE container to obtain the container's activity information:

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/monitor/MonitorActivityTrackerImpl.java#L104C57-L121

It seems that this is the principle behind the Operator monitoring the state of the IDE container and destroying the session to terminate the IDE. 😄

**Reply by inlann** [reply]
I found that the value of timeoutAfter, used in the pingSession function, is obtained from appDefinition, and this appDefinition is defined in theia-cloud-helm:

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/monitor/MonitorActivityTrackerImpl.java#L83C15-L88

https://github.com/eclipsesource/theia-cloud-helm/blob/4b0260fc8c80b9d2b15d087a232319b2a8f42e64/charts/theia-cloud-crds/templates/appdefinition-spec-resource.yaml#L75-L86

https://github.com/eclipsesource/theia-cloud-helm/blob/4b0260fc8c80b9d2b15d087a232319b2a8f42e64/charts/theia.cloud/values.yaml#L308-L314

The `appdefinition-spec-resource.yaml` file contains multiple versions ranging from v1beta to v8beta:

https://github.com/eclipsesource/theia-cloud-helm/blob/4b0260fc8c80b9d2b15d087a232319b2a8f42e64/charts/theia-cloud-crds/templates/appdefinition-spec-resource.yaml#L13-L491

And in the `appdefinition.java` within the operator, there is a @Version("v8beta") decorator:

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/java/common/org.eclipse.theia.cloud.common/src/main/java/org/eclipse/theia/cloud/common/k8s/resource/AppDefinition.java#L28-L32

However, in my cluster, the actual appdefinition CR (Custom Resource) displays the version as v1beta, which lacks the monitor field:

```bash
$ kubectl -n theiacloud describe appdefinitions.theia.cloud theia-cloud-demo 
Name:         theia-cloud-demo
Namespace:    theiacloud
Labels:       app.kubernetes.io/managed-by=Helm
Annotations:  meta.helm.sh/release-name: theia-cloud
              meta.helm.sh/release-namespace: theiacloud
API Version:  theia.cloud/v1beta
Kind:         AppDefinition
Metadata:
  Creation Timestamp:  2023-11-08T10:07:43Z
  Generation:          5
  Resource Version:    5478287
  UID:                 0e8583e3-9cd5-409f-877d-ac8f76a1dbd9
Spec:
  Downlink Limit:   30000
  Image:            theiacloud/theia-cloud-demo:0.8.1.MS3
  Ingressname:      theia-cloud-demo-ws-ingress
  Limits Cpu:       2
  Limits Memory:    1200M
  Max Instances:    10
  Min Instances:    0
  Name:             theia-cloud-demo
  Port:             3000
  Requests Cpu:     100m
  Requests Memory:  1000M
  UID:              101
  Uplink Limit:     30000
Events:             <none>
```

This absence prevents the operator from terminating the IDE container.

My other question is, how is the actual appdefinition CR created, and based on what does it set its version? Additionally, I wonder if directly modifying its fields using kubectl edit -n theia-cloud appdefinition would reset its version.

**Reply by sgraband** [reply, marked as answer]
Your assumption is correct, that the operator is the one who is pinging those endpoints and then decides when to shut down the pods.

You always get the `v1beta` field with kubectl because we currently have invalid version numbers (e.g. `v8beta`) and it will always default to `v1beta`. Note that the resource is still available as `v8beta` on the cluster, e.g. by changing your command to `kubectl -n theiacloud describe appdefinitions.v8beta.theia.cloud theia-cloud-demo`. We are planning to soonish update the version numbers to valid numbers (e.g. `v1beta8`). Then the default output of kubectl should be the latest version. Sorry for the inconvience.

**Comment by inlann** [comment]
BTW: I don't find the monitor-theia extension in Theia-Cloud official cluster:

![img_v3_0263_94547040-e6aa-4996-af01-945d5705fe7g](https://github.com/eclipsesource/theia-cloud/assets/33852855/09774352-284b-479d-a44e-5b389caa4fe1)

or the vscode monitor extension:

![img_v3_0263_23e05856-c9fa-471a-9f46-627b1acbfc6g](https://github.com/eclipsesource/theia-cloud/assets/33852855/5a43479c-8a56-46d6-93be-ebdac0bc8d74)

Does this means that Theia-Cloud's official deployment does not use these two extensions?

The versions of theia/core and theia/plugin in the dependencies of this plugin are still 1.34.0. If I want to use this plugin, do its dependency versions need to be updated along with Theia's own version?

https://github.com/eclipsesource/theia-cloud/blob/59ee2028bef1dcf0bf34724ad2433a07b748edd9/node/monitor-theia/package.json#L11-L14

**Reply by sgraband** [reply]
The "official deployment" does not use the extension as it is limited to 30 minuted per pod anyway and we don't see a reason to use the monitoring there.

In theory the Theia extension should still work with newer versions. However, we plan to update this according to the [CR cycle of Theia](https://theia-ide.org/releases/), when we have time.
